src/index/base.cpp

```diff
@@ -414,10 +414,18 @@ IndexSummary BaseIndex::GetSummary() const
 void BaseIndex::SetBestBlockIndex(const CBlockIndex* block) {
     assert(!node::fPruneMode || AllowPrune());
 
-    m_best_block_index = block;
     if (AllowPrune() && block) {
         node::PruneLockInfo prune_lock;
         prune_lock.height_first = block->nHeight;
         WITH_LOCK(::cs_main, m_chainstate->m_blockman.UpdatePruneLock(GetName(), prune_lock));
     }
+
+    // Intentionally set m_best_block_index as the last step in this function,
```

Comment:
Maybe it could be helpful to additionally add a similar comment at the end of `BaseIndex::BlockConnected`, so that no one attempts to add references to *this after the call to `SetBestBlockIndex` there either, for the same reason.